8284900: Check InitialHeapSize and container memory limits before startup

xpbob opened the pull request:
Check InitialHeapSize and container memory limits before startup

```
Operating System Metrics:
    Provider: cgroupv1
    Effective CPU Count: 8
    CPU Period: 100000us
    CPU Quota: -1
    CPU Shares: -1
    List of Processors, 8 total: 
    0 1 2 3 4 5 6 7 
    List of Effective Processors: N/A
    List of Memory Nodes, 1 total: 
    0 
    List of Available Memory Nodes: N/A
    Memory Limit: 50.00M
    Memory Soft Limit: Unlimited
    Memory & Swap Limit: 60.00M
    Maximum Processes Limit: 4194305
```
command:-Xms60m -XshowSettings:system -version 




When the memory limit is exceeded, the startup fail with message


```
Error occurred during initialization of VM
Initial heap size set to a larger value than the container memory & swap limit
```

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8284900](https://bugs.openjdk.java.net/browse/JDK-8284900): Check InitialHeapSize and container memory limits before startup


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8256/head:pull/8256` \
`$ git checkout pull/8256`

Update a local copy of the PR: \
`$ git checkout pull/8256` \
`$ git pull https://git.openjdk.java.net/jdk pull/8256/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8256`

View PR using the GUI difftool: \
`$ git pr show -t 8256`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8256.diff">https://git.openjdk.java.net/jdk/pull/8256.diff</a>

</details>

DamonFool commented on src/hotspot/share/gc/shared/gcArguments.cpp:
```diff
@@ -127,6 +127,10 @@ void GCArguments::initialize_heap_flags_and_sizes() {
     }
   }
 
+  if (FLAG_IS_CMDLINE(InitialHeapSize) && InitialHeapSize >= os::physical_memory()) {
```
What will happen if the swap space is unlimited?
```
    Memory Soft Limit: Unlimited
    Memory & Swap Limit: Unlimited
```

xpbob replied:
> What will happen if the swap space is unlimited?
> 
> ```
>     Memory Soft Limit: Unlimited
>     Memory & Swap Limit: Unlimited
> ```

Thanks,If the swap space of the physical machine is enabled, it can run
Use Memory & Swap limits to Limit programs

iklam commented on src/hotspot/share/gc/shared/gcArguments.cpp:
```diff
@@ -126,7 +129,12 @@ void GCArguments::initialize_heap_flags_and_sizes() {
+      FLAG_IS_CMDLINE(InitialHeapSize) && (InitialHeapSize >= (julong) OSContainer::memory_and_swap_limit_in_bytes())) {
+    vm_exit_during_initialization("Initial heap size set to a larger value than the container memory & swap limit");
+  }
+  #endif
```
`OSContainer::memory_and_swap_limit_in_bytes()` is an expensive operation. It opens files and read their contents.  The returned value should be stored in a local to avoid calling this function twice.

iklam commented:
Could you update the bug report to indicate the exact command used? This is not very clear:

```
Start the command：cgexec -g memory:foo -Xms60m xxx
```

xpbob commented:
> Could you update the bug report to indicate the exact command used? This is not very clear:
> 
> ```
> Start the command：cgexec -g memory:foo -Xms60m xxx
> ```

@iklam 
Thanks to review.
I updated the simulation
```
Simulation environment：
cgexec -g memory:foo java -XshowSettings:system -Xms60m -version （The successful running）
use -XX:+AlwaysPreTouch to simulate a long-running program
cgexec -g memory:foo java -XshowSettings:system -Xms60m -XX:+AlwaysPreTouch -version(killed by cgroup) 
```

iklam commented on src/hotspot/share/gc/shared/gcArguments.cpp:
```diff
@@ -130,9 +130,11 @@ void GCArguments::initialize_heap_flags_and_sizes() {
+    jlong memswBytes = OSContainer::memory_and_swap_limit_in_bytes();
+    if ((memswBytes > 0) && FLAG_IS_CMDLINE(InitialHeapSize) && (InitialHeapSize >= (julong) memswBytes)) {
+      vm_exit_during_initialization("Initial heap size set to a larger value than the container memory & swap limit");
+    }
```
I noticed one more problem. InitialHeapSize is usually not set in the command-line. So the condition should be changed to the following to avoid unnecessary calls.

```
if (OSContainer::is_containerized() && FLAG_IS_CMDLINE(InitialHeapSize)) { ...
```

DamonFool commented on src/hotspot/share/gc/shared/gcArguments.cpp:
```diff
@@ -126,7 +129,14 @@ void GCArguments::initialize_heap_flags_and_sizes() {
     }
   }
-
+  #if defined(LINUX)
```
No space before `#if`.

DamonFool commented on src/hotspot/share/gc/shared/gcArguments.cpp:
```diff
@@ -126,7 +129,14 @@ void GCArguments::initialize_heap_flags_and_sizes() {
+      vm_exit_during_initialization("Initial heap size set to a larger value than the container memory & swap limit");
+    }
+  }
+  #endif
```
No space before `#endif`.

xpbob commented:
@iklam @DamonFool 
Thanks for review.
The code has been updated

stefank commented on src/hotspot/share/gc/shared/gcArguments.cpp:
```diff
@@ -126,7 +129,14 @@ void GCArguments::initialize_heap_flags_and_sizes() {
     }
   }
-
+#if defined(LINUX)
+  if (OSContainer::is_containerized() && FLAG_IS_CMDLINE(InitialHeapSize)) {
+    jlong memswBytes = OSContainer::memory_and_swap_limit_in_bytes();
+    if ((memswBytes > 0) && (InitialHeapSize >= (julong) memswBytes)) {
+      vm_exit_during_initialization("Initial heap size set to a larger value than the container memory & swap limit");
+    }
+  }
+#endif
```
The current code is structured with carefully added newlines to aid readability. This patch removes one of those newlines, and replace it with this block. Could you keep the current format and add code as a separate block with newlines before and after? Maybe also consider moving this to after the more primitive checks in 140-149?

xpbob commented:
@stefank 
Thanks for review.
The code has been updated

stefank reviewed (commented):
Thanks for the update. I'll let those that understand the OSContainer APIs review if this is the correct usage of it.

Small nit: Local variables in GC code tend to be written with underscores and not camelCase. Maybe just rename `memswBytes` to `limit`. A short name seems fine for such a small scope.

xpbob commented:
@stefank
Thanks for review.
The code has been updated

iklam commented:
I have updated the bug report [JDK-8284900](https://bugs.openjdk.java.net/browse/JDK-8284900), but there are still some details that are missing. I.e., how to set up the `foo` control group. @xpbob could you add the details into the bug report?

The bug report should have a concrete reproducible test case so anyone reading it in the future won't need to guess the reproduction steps.

iklam reviewed (changes requested):
Please add a new test case under https://github.com/openjdk/jdk/blob/0f81d8fcc3fb703760b1cddb01861ea5031023fb/test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java

xpbob commented:
@iklam 
Thanks for review.
I have updated the steps for creating a cgroup in the the bug report [JDK-8284900](https://bugs.openjdk.java.net/browse/JDK-8284900) and added test case

iklam commented on src/hotspot/share/gc/shared/gcArguments.cpp:
```diff
@@ -138,6 +141,15 @@ void GCArguments::initialize_heap_flags_and_sizes() {
+    jlong limit = OSContainer::memory_and_swap_limit_in_bytes();
+    if ((limit > 0) && (InitialHeapSize >= (julong) limit)) {
+      vm_exit_during_initialization("Initial heap size set to a larger value than the container memory & swap limit");
+    }
```
Since you are testing `InitialHeapSize >= (julong) limit)`, the error message should be:

> Initial heap size should be smaller than the container memory & swap limit

The JVM will use memory outside of the heap (for example malloc, or code cache), so even if you set the `InitialHeapSize` to below the memory limit, the JVM might still get killed by OOM unpredictably.

This PR catches obvious problems when `-Xms` is set way too large. However, I think we need to consider the larger issue of OOM killer when using Java in containers. I filed [JDK-8285277](https://bugs.openjdk.java.net/browse/JDK-8285277)

I've started a discussion on [hotspot-dev](https://mail.openjdk.java.net/pipermail/hotspot-dev/2022-April/059463.html) about what approach we should use to handle container memory limits. Let's hold on with this PR until we have a consensus.

xpbob replied:
> Since you are testing `InitialHeapSize >= (julong) limit)`, the error message should be:
> 
> > Initial heap size should be smaller than the container memory & swap limit
> 
> The JVM will use memory outside of the heap (for example malloc, or code cache), so even if you set the `InitialHeapSize` to below the memory limit, the JVM might still get killed by OOM unpredictably.
> 
> This PR catches obvious problems when `-Xms` is set way too large. However, I think we need to consider the larger issue of OOM killer when using Java in containers. I filed [JDK-8285277](https://bugs.openjdk.java.net/browse/JDK-8285277)
> 
> I've started a discussion on [hotspot-dev](https://mail.openjdk.java.net/pipermail/hotspot-dev/2022-April/059463.html) about what approach we should use to handle container memory limits. Let's hold on with this PR until we have a consensus.

Thanks
The code I want to achieve here is similar to the physical machine situation, with only Xms memory guaranteed

jerboaa commented on src/hotspot/share/gc/shared/gcArguments.cpp:
```diff
@@ -138,6 +141,15 @@ void GCArguments::initialize_heap_flags_and_sizes() {
   }
 
+#if defined(LINUX)
+  if (OSContainer::is_containerized() && FLAG_IS_CMDLINE(InitialHeapSize)) {
+    jlong limit = OSContainer::memory_and_swap_limit_in_bytes();
+    if ((limit > 0) && (InitialHeapSize >= (julong) limit)) {
```
Perhaps we should generalize this to use `os::physical_memory()` here instead of the Linux-specific `OSContainer` API? `os::physical_memory()` would return the container limit if it detects so.

DamonFool replied:
> Perhaps we should generalize this to use `os::physical_memory()` here instead of the Linux-specific `OSContainer` API? `os::physical_memory()` would return the container limit if it detects so.

What will happen with `-Xms7G` on a machine configured with 4G physical memory + 4G swap space?
Does `os::physical_memory()`  return `physical memory + swap space size`?

stefank replied:
@jerboaa Good point. Also take a look at Arguments::set_heap_size() and how that function uses os::physical_memory() and also limit_heap_by_allocatable_memory(). It seems like any change in GCArguments should go hand-in-hand with the code in set_heap_size.

jerboaa replied:
> > Perhaps we should generalize this to use `os::physical_memory()` here instead of the Linux-specific `OSContainer` API? `os::physical_memory()` would return the container limit if it detects so.
> 
> What will happen with `-Xms7G` on a machine configured with 4G physical memory + 4G swap space? Does `os::physical_memory()` return `physical memory + swap space size`?

It doesn't. What would happen on a physical machine, though? It should really do the same thing regardless of the container settings. Container settings, in a nutshell, should be an abstraction of "physical" host. Otherwise behave the same. Having said that, in containers (at least kubernetes) swap is usually disabled. Also, does it make sense to allow `-Xms7g` on a system with `4G` of physical RAM? I'm doubtful.

jerboaa replied:
> @jerboaa Good point. Also take a look at Arguments::set_heap_size() and how that function uses os::physical_memory() and also limit_heap_by_allocatable_memory(). It seems like any change in GCArguments should go hand-in-hand with the code in set_heap_size.

+1

dholmes-ora replied:
I agree with the comments that this should behave exactly as it would for a physical machine. The fact we are in a container is not really relevant.

iklam replied:
Maybe we ought to move the check into arguments.cpp, and compare os::physical_memory against (initial heap size + initial code cache size + .... something else)?

Forcing the VM to quit may have backward compatibility issues, though. Some apps may set their -Xms to be larger than the container memory limit, but they may use so little heap memory that the VM never commits enough memory to trigger the OOM killer. With this PR those apps will suddenly start failing.

xpbob commented:
Thanks for review.
@iklam @jerboaa @DamonFool @stefank @dholmes-ora 

os::physical_memory() return memory.limit_in_bytes in cgroupv1. In a container environment, swap is usually disabled, it makes sense to limit xms with this. The detection here needs to work only under the container.

The JVM is currently unable to calculate the memory footprint of a program at startup.Use the memory of the user's thread, which is available after startup.

Currently Xms Settings in non-container environments exceed physical machine memory and exit,The container should behave like a physical machine
```
java -Xms40g -version

penJDK 64-Bit Server VM warning: INFO: os::commit_memory(0x00007fd8ac600000, 28633464832, 0) failed; error='Cannot allocate memory' (errno=12)
#
# There is insufficient memory for the Java Runtime Environment to continue.
# Native memory allocation (mmap) failed to map 28633464832 bytes for committing reserved memory.
# An error report file with more information is saved as:
# /data/test/docker/hs_err_pid15626.log
```

xpbob commented:
so any suggestions?

brunoborges commented:
@jerboaa and folks, is this PR still in progress or is attention moving towards the lager problem?
